### PR TITLE
Changes to allow external AuthZ providers to integrate

### PR DESCRIPTION
**Pull request by aayustark007-fk**
Separate out the Authorization interfaces to separate module.
Add maven publish plugin to allow uploading artefacts to maven.
Before publishing add the remote repo url to `gradle.properties` for `mavenPublishUrl`

### CODE REVIEW

File: authz/src/test/java/com/flipkart/varadhi/authz/DefaultAuthorizationProviderTest.java [hunk 1]
```diff
@@ -197,28 +198,42 @@ public void testIsAuthorized_UserProjectAccess(VertxTestContext testContext) {
...
+                .compose(t -> defaultAuthorizationProvider.isAuthorized(testUser("proj_user1", false),
                         ResourceAction.PROJECT_GET, "flipkart/team_rocket/proj001"
                 )) // proj_user1 is project.read so should work
                 .onComplete(testContext.succeeding(t -> {
                     Assertions.assertTrue(t);
                     checkpoint.flag();
                 }));
     }
+
+    public static UserContext testUser(String name, boolean expired) {
```
Comment (kmrdhruv) [inline authz/src/test/java/com/flipkart/varadhi/authz/DefaultAuthorizationProviderTest.java:226]: Might be better to put into testFixtures, can possibly used by other modules as well ?

File: buildSrc/src/main/groovy/com.flipkart.varadhi.java-library-conventions.gradle [hunk 2]
```diff
@@ -4,4 +4,29 @@ plugins {
...
+}
+
+
+publishing {
+    publications {
+        maven(MavenPublication) {
+            groupId = "${projectGroupId}"
+            artifactId = "${project.name}"
+            version = "${version}"
+            from components.java
```
Comment (kmrdhruv) [inline buildSrc/src/main/groovy/com.flipkart.varadhi.java-library-conventions.gradle:19]: What all artifects/components will get published with this ?

File: buildSrc/src/main/groovy/com.flipkart.varadhi.java-library-conventions.gradle [hunk 3]
```diff
@@ -4,4 +4,29 @@ plugins {
...
+    id 'maven-publish'
+}
+
+
+publishing {
+    publications {
+        maven(MavenPublication) {
+            groupId = "${projectGroupId}"
+            artifactId = "${project.name}"
+            version = "${version}"
```
Comment (kmrdhruv) [inline buildSrc/src/main/groovy/com.flipkart.varadhi.java-library-conventions.gradle:18]: How will this result in snapshot or unstable version being published ?

File: buildSrc/src/main/groovy/com.flipkart.varadhi.java-library-conventions.gradle [hunk 4]
```diff
@@ -4,4 +4,29 @@ plugins {
 
     // Apply the java-library plugin for API and implementation separation.
     id 'java-library'
+
+    // Apply the maven-publish to allow deploying to maven artifactories
+    id 'maven-publish'
+}
+
+
+publishing {
```
Comment (kmrdhruv) [inline buildSrc/src/main/groovy/com.flipkart.varadhi.java-library-conventions.gradle:13]: This will add publishing task to each subproject ? Not all libraries needs to be published ?

File: buildSrc/src/main/groovy/com.flipkart.varadhi.java-library-conventions.gradle [hunk 5]
```diff
@@ -4,4 +4,29 @@ plugins {
...
+
+            pom {
+                name = "${project.name}"
+            }
+        }
+    }
+
+    repositories {
+        maven {
+            url = mavenPublishUrl.replace('"', '') // important to strip double-quotes from this otherwise the url will be dealt as a file url
```
Comment (kmrdhruv) [inline buildSrc/src/main/groovy/com.flipkart.varadhi.java-library-conventions.gradle:29]: yes, let's get this fixed as well.

File: buildSrc/src/main/groovy/com.flipkart.varadhi.java-library-conventions.gradle [hunk 6]
```diff
@@ -4,4 +4,29 @@ plugins {
...
+publishing {
+    publications {
+        maven(MavenPublication) {
+            groupId = "${projectGroupId}"
+            artifactId = "${project.name}"
+            version = "${version}"
+            from components.java
+
+            pom {
+                name = "${project.name}"
```
Comment (kmrdhruv) [inline buildSrc/src/main/groovy/com.flipkart.varadhi.java-library-conventions.gradle:22]: Do we need other details as well here ?

File: authz/build.gradle [hunk 7]
```diff
@@ -0,0 +1,17 @@
+plugins {
```
Comment (kmrdhruv) [inline authz/build.gradle:1]: would it make sense to move entities/auth as well into this module ?

**Review by kmrdhruv — CHANGES_REQUESTED** [review]

File: authz/build.gradle (hunk 7 above)
Comment (aayustark007-fk) [reply authz/build.gradle:1]: if modules are owning their entities then it would make the entities module redundant

File: authz/src/test/java/com/flipkart/varadhi/authz/DefaultAuthorizationProviderTest.java (hunk 1 above)
Comment (aayustark007-fk) [reply authz/src/test/java/com/flipkart/varadhi/authz/DefaultAuthorizationProviderTest.java:226]: sure will do that

File: buildSrc/src/main/groovy/com.flipkart.varadhi.java-library-conventions.gradle (hunk 4 above)
Comment (aayustark007-fk) [reply buildSrc/src/main/groovy/com.flipkart.varadhi.java-library-conventions.gradle:13]: any module using java-library-conventions will get published
shall we can create a separate convention extending this for publishable-libraray-convention?

File: buildSrc/src/main/groovy/com.flipkart.varadhi.java-library-conventions.gradle (hunk 3 above)
Comment (aayustark007-fk) [reply buildSrc/src/main/groovy/com.flipkart.varadhi.java-library-conventions.gradle:18]: won't be published unless gradlew publish is run
I have published SNAPSHOT version, but pushed changes have omitted SNAPSHOT since the docker build was failing

File: buildSrc/src/main/groovy/com.flipkart.varadhi.java-library-conventions.gradle (hunk 2 above)
Comment (aayustark007-fk) [reply buildSrc/src/main/groovy/com.flipkart.varadhi.java-library-conventions.gradle:19]: authz, commons and entities

File: buildSrc/src/main/groovy/com.flipkart.varadhi.java-library-conventions.gradle (hunk 6 above)
Comment (aayustark007-fk) [reply buildSrc/src/main/groovy/com.flipkart.varadhi.java-library-conventions.gradle:22]: this is to generate pom.xml so that the jar plays well when including in maven projects

Comment (aayustark007-fk) [reply buildSrc/src/main/groovy/com.flipkart.varadhi.java-library-conventions.gradle:22]: we can add more details if required, but currently I don't think anything else is required

File: buildSrc/src/main/groovy/com.flipkart.varadhi.java-library-conventions.gradle (hunk 5 above)
Comment (aayustark007-fk) [reply buildSrc/src/main/groovy/com.flipkart.varadhi.java-library-conventions.gradle]: done

File: authz/src/test/java/com/flipkart/varadhi/authz/DefaultAuthorizationProviderTest.java (hunk 1 above)
Comment (aayustark007-fk) [reply authz/src/test/java/com/flipkart/varadhi/authz/DefaultAuthorizationProviderTest.java]: done

File: authz/build.gradle (hunk 7 above)
Comment (kmrdhruv) [reply authz/build.gradle:1]: agree for Varadhi domain entities, but entities/auth has enum like resource types and resource actions and config which seems different from other entities. It has no intended uses in other modules (except authz and server).

File: buildSrc/src/main/groovy/com.flipkart.varadhi.java-library-conventions.gradle (hunk 4 above)
Comment (kmrdhruv) [reply buildSrc/src/main/groovy/com.flipkart.varadhi.java-library-conventions.gradle:13]: Yes, either separate publishing convention or add it specifically to the required modules. May be former is better.

File: buildSrc/src/main/groovy/com.flipkart.varadhi.java-library-conventions.gradle (hunk 3 above)
Comment (kmrdhruv) [reply buildSrc/src/main/groovy/com.flipkart.varadhi.java-library-conventions.gradle:18]: I think in code it should be SNAPSHOT by default. Why docker build was failing ?
(what I meant, is adding publishing to common convention will make that target visible to all submodules. Someone running publish at root, will execute it for all modules which we can avoid).

File: buildSrc/src/main/groovy/com.flipkart.varadhi.java-library-conventions.gradle (hunk 2 above)
Comment (kmrdhruv) [reply buildSrc/src/main/groovy/com.flipkart.varadhi.java-library-conventions.gradle:19]: What I was asking is w.r.to  "components.java" ? What all does it publish for a single library module.

**Review by kmrdhruv — CHANGES_REQUESTED** [review]
added few comments for clarification.

File: buildSrc/src/main/groovy/com.flipkart.varadhi.java-library-conventions.gradle (hunk 3 above)
Comment (aayustark007-fk) [reply buildSrc/src/main/groovy/com.flipkart.varadhi.java-library-conventions.gradle]: docker build is hardcoded to use specific verison
I'll fix that

File: buildSrc/src/main/groovy/com.flipkart.varadhi.java-library-conventions.gradle (hunk 2 above)
Comment (aayustark007-fk) [reply buildSrc/src/main/groovy/com.flipkart.varadhi.java-library-conventions.gradle]: not sure about this, I saw it was provided as part of publish syntax
published jar has to be maven compatible so it adds the required metadata to be able to correctly indexed by maven repos

File: buildSrc/src/main/groovy/com.flipkart.varadhi.java-library-conventions.gradle (hunk 4 above)
Comment (aayustark007-fk) [reply buildSrc/src/main/groovy/com.flipkart.varadhi.java-library-conventions.gradle]: created separate publish config

File: authz/build.gradle (hunk 7 above)
Comment (aayustark007-fk) [reply authz/build.gradle:1]: done

File: setup/docker/Dockerfile [hunk 8]
```diff
@@ -14,7 +14,7 @@ EXPOSE 8080
 RUN mkdir -p /usr/share/varadhi/dependencies && chmod 755 /usr/share/varadhi/dependencies
 RUN mkdir /etc/varadhi && chmod 755 /etc/varadhi
 COPY $VARADHI_HOME/server/build/dependencies/ /usr/share/varadhi/dependencies/
-COPY $VARADHI_HOME/pulsar/build/libs/pulsar-0.1.0.1.jar /usr/share/varadhi/dependencies/
+COPY $VARADHI_HOME/pulsar/build/libs/pulsar-*.jar /usr/share/varadhi/dependencies/
```
Comment (kmrdhruv) [inline setup/docker/Dockerfile:17]: Why is this needed ?

**Review by kmrdhruv — APPROVED** [review]
Looks fine. Let's verify what all gets published. Not sure, iirc, it had lot more artifacts as part of publish, see if it is required.
